### ISSUE TITLE
Pull Request create Event triggered too early

### ISSUE THREAD

**Issue by patst** [Eernie/bitbucket-webhooks-plugin#50]
We use the bitbucket-webhooks-plugin to trigger jenkins builds for opened pull requests.
After the latest plugin updates inside jenkins the first build for newly created pull requests fails during Jenkinsfile checkout:
``` > git fetch --no-tags --progress https://<server>/scm/<project>/<project>.git +refs/pull-requests/102/from:refs/remotes/origin/PR-102
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Command "git fetch --no-tags --progress https://<server>/scm/<repo>/<project>.git +refs/pull-requests/102/from:refs/remotes/origin/PR-102" returned status code 128:
stdout: 
stderr: fatal: Couldn't find remote ref refs/pull-requests/102/from

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1903)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1622)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:71)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:348)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:545)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1075)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1115)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:113)
	at org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition.create(CpsScmFlowDefinition.java:130)
	at org.jenkinsci.plugins.workflow.multibranch.SCMBinder.create(SCMBinder.java:120)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:262)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:405)
[Bitbucket] Notifying pull request build result
[Bitbucket] Build result notified
ERROR: Error cloning remote repo 'origin'
Finished: FAILURE
```
If we restart the build in jenkins it succeeds. The same is true for changes to pull requests.

It seems the internal bitbucket structure is not ready at the time the webhook is triggered. 
I read this article: https://answers.atlassian.com/questions/239988 (I found the link in `PullRequestListener`)

The first build succeeds if I insert a `Thread.sleep(1000)` before sending the webhooks POST request.
It also succeds if I call the `canMerge(.....)` method when the `PullRequestOpenedEvent` is fired.

**Comment by casz** [comment]
I have been seeing the same issue.

**Comment by dantran** [comment]
is it a regression due to major changes at lastest bitbucket branch source plugin from @stephenc

**Comment by stephenc** [comment]
We consulted with @HRMPW @i386 and @erikvanzijst and the view was that the Jenkins plugin should use the `refs/pull-requests/*/from` rather than trying to pull the ref from the upstream repository.

A side effect of that switch is that it now exposes a bug in how Bitbucket fails to update the refs consistently. After a discussion with @HRMPW the decision was that this was a clear bug in Bitbucket and that Jenkins should not work around bugs in Bitbucket (the only option available to Jenkins was to try and make a HTTP request to the PR's HTML page, e.g. `https://bitbucket.exmaple.com/projects/${key}/repos/${repo}/pull-requests/${prNum}/overview`) 

This PR change seems much less hacky (though I think there should be a `// TODO remove once Bitbucket fixes bug`)

I would also be inclined to have a comment explaining why the call to can merge, the link to https://community.atlassian.com/t5/Bitbucket-questions/Re-Change-pull-request-refs-after-Commit-instead-of-after/qaq-p/194703/comment-id/6829#M6829 is probably sufficient

FTR we want to move the BB cloud code to use the same style of refs... once BB cloud gets https://bitbucket.org/site/master/issues/5814/reify-pull-requests-by-making-them-a-ref

**Comment by patst** [comment]
Thanks for the feedback. I added a TODO for removal and a link for the explanation what canMerge does.

**Comment by casz** [comment]
Fixed in 1.5.0
